File: 187.py
from collections import defaultdict
from typing import List


# class Solution:
#     def findRepeatedDnaSequences(self, s: str) -> List[str]:
#         L = 10
#         bin = {'A': 0, 'C': 1, 'G': 2, 'T': 3}
#         n = len(s)
#         if n <= L:
#             return []
#         res = []
#         key = 0
#         dic = defaultdict(int)
#         for ch in s[:L - 1]:
#             key = (key << 2) | bin[ch]
#         for index in range(L - 1, n):
#             key = ((key << 2) | bin[s[index]]) & ((1 << L * 2) - 1)
#             dic[key] += 1
#             if dic[key] == 2:
#                 res.append(s[index - L + 1 : index + 1])
#         return res

class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        MOD = 10**9 + 7
        n = len(s)
        P = 131313
        h = [0] * n
        p = [0] * n
        p[0] = 1
        ans = []
        for i in range(1, n):
            h[i] = h[i - 1] * P + s[i-1]
            p[i] = p[i - 1] * P
        dic = defaultdict(int)
        for i in range(n-10):
            j = i + 9
            hash = h[j] - h[i - 1] * p[j - i + 1]
            cnt = dic[hash]
            if cnt == 1: ans.append(s[i:j+1])
            dic[hash] += 1
        return ans
           


            

solu = Solution()
print(solu.findRepeatedDnaSequences("AAAAAAAAAAAAA"))

File: go/187.go
package main

func findRepeatedDnaSequences(s string) (ans []string) {
	N := 10
	m := make(map[string]int)
	for i := 0; i < len(s)-N; i++ {
		subS := s[i : i+10]
		m[subS] += 1
		if m[subS] == 2 {
			ans = append(ans, subS)
		}
	}
	return ans
}

func findRepeatedDnaSequences2(s string) (ans []string) {
	N := 10
	if len(s) < N+1 {
		return ans
	}
	bin := map[byte]int{'A': 0, 'C': 1, 'G': 2, 'T': 3}
	m := make(map[int]int)
	key := 0
	for i := 0; i < N-1; i++ {
		key = (key << 2) | bin[s[i]]
	}
	for i := N - 1; i < len(s); i++ {
		key = ((key << 2) | bin[s[i]]) & (1<<(2*N) - 1)
		m[key] += 1
		if m[key] == 2 {
			ans = append(ans, s[i-N+1:i+1])
		}
	}
	return ans
}

// func main() {

// }
